### PR TITLE
Use windows registry to discover python

### LINKED ISSUE #1521: PEP 514 support
https://peps.python.org/pep-0514/

We use `py` to find Python installations on Windows today but we want to switch to registry look-ups per PEP 514

### LINKED ISSUE #6524: `uv python list` fails if a pyenv-win install is included in the windows registry
If a version of python is installed via pyenv with the `--register` flag then subsequently `uv` will fail to list python installs as it attempts to query metadata of a file that does not exist.

It looks like this is caused because `uv` makes the assumption that `py --list-paths` will return paths with only `<major>.<minor>` version information while pyenv installs are shown with `<major>.<minor>.<micro>`[^1]. The regex is then capturing the `<micro>` part of the version as part of the path to the python executable.

https://github.com/astral-sh/uv/blob/bbcd10d3cc7710b8c33955b91915e8463c2b6b9b/crates/uv-python/src/py_launcher.rs#L36-L39

Commands:
```
pyenv install 3.12.5 --register
uv python list
```

Full output:

```
C:\>uv --version
uv 0.3.2 (c5440001c 2024-08-23)

C:\>py --list
No installed Pythons found!

C:\>uv python list
cpython-3.12.5-windows-x86_64-none     <download available>
cpython-3.11.9-windows-x86_64-none     <download available>
cpython-3.10.14-windows-x86_64-none    <download available>
cpython-3.9.19-windows-x86_64-none     <download available>
cpython-3.8.19-windows-x86_64-none     <download available>
pypy-3.7.13-windows-x86_64-none        <download available>

C:\>pyenv install 3.12.5 --register
:: [Info] ::  Mirror: https://www.python.org/ftp/python
:: [Installing] ::  3.12.5 ...
:: [Info] :: completed! 3.12.5

C:\>py --list-paths
 -V:3.12.5 *      C:\Users\ducks\.pyenv\pyenv-win\versions\3.12.5\python.exe

C:\>uv python list
error: Failed to query Python interpreter
  Caused by: failed to query metadata of file `C:\.5 *      C:\Users\ducks\.pyenv\pyenv-win\versions\3.12.5\python.exe`
  Caused by: The filename, directory name, or volume label syntax is incorrect. (os error 123)
```

Example of `py --list-paths` output with multiple versions of python installed and registered with pyenv and some regular installs.[^2]

```
 -V:3.13.0rc1 *   C:\Users\ducks\.pyenv\pyenv-win\versions\3.13.0rc1\python.exe
 -V:3.12.5        C:\Users\ducks\.pyenv\pyenv-win\versions\3.12.5\python.exe
 -V:3.12.5-win32  C:\Users\ducks\.pyenv\pyenv-win\versions\3.12.5-win32\python.exe
 -V:3.12          C:\Users\ducks\AppData\Local\Programs\Python\Python312\python.exe
 -V:3.12-32       C:\Users\ducks\AppData\Local\Programs\Python\Python312-32\python.exe
 -V:3.11.9        C:\Users\ducks\.pyenv\pyenv-win\versions\3.11.9\python.exe
 -V:3.10.11       C:\Users\ducks\.pyenv\pyenv-win\versions\3.10.11\python.exe
 -V:3.9.13        C:\Users\ducks\.pyenv\pyenv-win\versions\3.9.13\python.exe
 -V:3.8.10        C:\Users\ducks\.pyenv\pyenv-win\versions\3.8.10\python.exe
```

[^1]: It looks like `py` is just using the key names to get the version rather than the value of `Version` or `SysVersion` in the registry.
[^2]: This also lead to me finding out that pyenv-win registers 32bit python installs as 64bit which I've reported as a bug there.

### PR DESCRIPTION
**Pull request by konstin**
Our current strategy of parsing the output of `py --list-paths` to get the installed python versions on windows is brittle (#6524, missing `py`, etc.) and it's slow (10ms last time i measured).

Instead, we should behave spec-compliant and read the python versions from the registry following PEP 514.

It's not fully clear which errors we should ignore and which ones we need to raise.

We're using the official rust-for-windows crates for accessing the registry.

Fixes #1521
Fixes #6524

### CODE REVIEW

File: crates/uv-python/src/discovery.rs [hunk 1]
```diff
@@ -311,15 +310,24 @@ fn python_executables_from_installed<'a>(
...
-                    .map(|entries|
-                // We can avoid querying the interpreter using versions from the py launcher output unless a patch is requested
-                entries.into_iter().filter(move |entry|
-                    version.is_none() || version.is_some_and(|version|
-                        version.has_patch() || version.matches_major_minor(entry.major, entry.minor)
-                    )
-                )
-                .map(|entry| (PythonSource::PyLauncher, entry.executable_path)))
+                registry_pythons()
+                    // We can avoid querying the interpreter using versions from the py launcher output unless a patch is requested
```
Comment (zanieb) [inline crates/uv-python/src/discovery.rs:314]: Comment needs to be updated.

File: crates/uv-python/src/py_launcher.rs [hunk 2]
```diff
@@ -1,93 +1,111 @@
...
-    // Find the first python of the version we want in the list
-    let stdout = String::from_utf8(output.stdout).map_err(|err| Error::StatusCode {
-        message: format!("The stdout of `py --list-paths` isn't UTF-8 encoded: {err}"),
-        exit_code: output.status,
-        stdout: String::from_utf8_lossy(err.as_bytes()).trim().to_string(),
-        stderr: String::from_utf8_lossy(&output.stderr).trim().to_string(),
-    })?;
+    // The registry has no natural ordering, so we're processing the latest version first.
+    registry_pythons.sort_by(|a, b| {
+        // Highest version first (reverse), but missing versions at the bottom (regular order).
```
Comment (zanieb) [inline crates/uv-python/src/py_launcher.rs:54]: What's a missing version? A registered Python without a version?

File: crates/uv-python/src/py_launcher.rs [hunk 3]
```diff
@@ -1,93 +1,111 @@
...
+fn read_registry_entry(company: &str, tag: &str, tag_key: &Key) -> Option<RegistryPython> {
+    // `ExecutablePath` is mandatory for executable Pythons.
+    let Some(executable_path) = tag_key
+        .open("InstallPath")
+        .and_then(|install_path| install_path.get_value("ExecutablePath"))
+        .ok()
+        .and_then(value_to_string)
+    else {
+        debug!(
+            r"Registry Python is not executable: `Software\Python\{}\{}",
```
Comment (zanieb) [inline crates/uv-python/src/py_launcher.rs:78]: Should we say like "Python interpreter in the registry is not executable"?

File: crates/uv-python/src/py_launcher.rs [hunk 4]
```diff
@@ -1,93 +1,111 @@
...
+            company, tag
+        );
+        return None;
+    };
+    debug!(
+        "Found registry Python {} {}: `{}`",
```
Comment (zanieb) [inline crates/uv-python/src/py_launcher.rs:84]: We do have a print for this in the discovery code

```rust
                debug!(
                    "Found `{}` at `{}` ({source})",
                    interpreter.key(),
                    path.display()
                );
```

Is this different? Do we need it?

File: crates/uv-python/src/py_launcher.rs [hunk 5]
```diff
@@ -1,93 +1,111 @@
...
+            Value::String(s) => Some(s),
+            _ => None,
+        })
+        .and_then(|s| match PythonVersion::from_str(&s) {
+            Ok(version) => Some(version),
+            Err(err) => {
+                debug!(
+                    "Invalid registry Python version at {}: {}",
+                    executable_path, err
+                );
```
Comment (zanieb) [inline crates/uv-python/src/py_launcher.rs:102]: Similar to above, I'd say like "Skipping Python interpreter ({}) with invalid registry version {s}: {err}"

We have a message like this elsewhere:

https://github.com/astral-sh/uv/blob/7502a963e15ae40422f8386b64f8b6c72af22d82/crates/uv-python/src/discovery.rs#L621

File: crates/uv-python/src/py_launcher.rs (hunk 2 above)
Comment (konstin) [reply crates/uv-python/src/py_launcher.rs:54]: A `None`

Comment (zanieb) [reply crates/uv-python/src/py_launcher.rs:54]: Makes sense, can you make that clear in the comment though? I didn't learn what we did with missing versions until further down.

File: crates/uv-python/src/py_launcher.rs [hunk 6]
```diff
@@ -1,93 +1,108 @@
...
+/// Find all Pythons registered in the Windows registry following PEP 514.
+pub(crate) fn registry_pythons() -> Result<Vec<RegistryPython>, windows_result::Error> {
+    let mut registry_pythons = Vec::new();
+    for root_key in [CURRENT_USER, LOCAL_MACHINE] {
+        let Ok(key_python) = root_key.open(r"Software\Python") else {
+            continue;
+        };
+        for company in key_python.keys()? {
+            // Reserved name according to the PEP.
+            if company == "Registry" {
```
Comment (zooba) [inline crates/uv-python/src/py_launcher.rs:35]: I think you mean `"PyLauncher"` here.

**Review by zooba — COMMENTED** [review]
As PEP 514 author, I'm happy to see this!

Do note that Store installs don't always show up properly in the registry due to how Windows handles its redirection (and it changes over time as Python chooses different redirection settings). For the `py.exe` launcher, we also do a [specific search](https://github.com/python/cpython/blob/61bef6245c4a32bf430d684ede8603f423d63284/PC/launcher2.c#L1744) for [known package IDs](https://github.com/python/cpython/blob/61bef6245c4a32bf430d684ede8603f423d63284/PC/launcher2.c#L1963) to locate Store installs. Not sure whether you want to copy that or not.

**Comment by zanieb** [comment]
Thank you for taking a look @zooba!

**Comment by konstin** [comment]
Thanks for the reminder!

File: docs/concepts/python-versions.md [hunk 7]
```diff
@@ -263,7 +263,8 @@ creating optimized, performant builds (e.g., with PGO and LTO enabled) is very s
 These distributions have some behavior quirks, generally as a consequence of portability; and, at
 present, uv does not support installing them on musl-based Linux distributions, like Alpine Linux.
 See the
-[`python-build-standalone` quirks](https://gregoryszorc.com/docs/python-build-standalone/main/quirks.html)
+[
+`python-build-standalone` quirks](https://gregoryszorc.com/docs/python-build-standalone/main/quirks.html)
```
Comment (zanieb) [inline docs/concepts/python-versions.md:267]: Looks wrong
```suggestion
[`python-build-standalone` quirks](https://gregoryszorc.com/docs/python-build-standalone/main/quirks.html)
```

Comment (konstin) [reply docs/concepts/python-versions.md]: I ran prettier now

**Pull request merged**